benchmarks/ddp/benchmark.py

```diff
@@ -0,0 +1,276 @@
+import time
+
+import numpy as np
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+import torch.optim as optim
+import torchvision
+
+
+def allgather_object(obj):
+    buffer = io.BytesIO()
+    torch.save(obj, buffer)
+    input_tensor = torch.ByteTensor(list(buffer.getvalue()))
+    input_length = torch.IntTensor([input_tensor.size(0)])
+    dist.all_reduce(input_length, op=dist.ReduceOp.MAX)
+    input_tensor.resize_(input_length[0])
+    output_tensors = [
+        torch.empty(input_tensor.size(), dtype=torch.uint8)
+        for _ in range(dist.get_world_size())
+    ]
+    dist.all_gather(output_tensors, input_tensor)
+    output = []
+    for tensor in output_tensors:
+        buffer = io.BytesIO(np.asarray(tensor).tobytes())
+        output.append(torch.load(buffer))
+    return output
+
+
+def allgather_run(cmd):
+    proc = subprocess.run(shlex.split(cmd), capture_output=True)
+    assert(proc.returncode == 0)
+    return allgather_object(proc.stdout.decode("utf-8"))
+
+
+
+def allequal(iterator):
+    iterator = iter(iterator)
+    try:
+        first = next(iterator)
+    except StopIteration:
+        return True
+    return all(first == rest for rest in iterator)
+
+
+def benchmark_process_group(pg, benchmark):
+    torch.manual_seed(pg.rank())
+    torch.cuda.manual_seed(pg.rank())
+
+    model = benchmark.create_model()
+    data = [(benchmark.generate_inputs(), benchmark.generate_target())]
+    criterion = nn.CrossEntropyLoss()
+    optimizer = optim.SGD(model.parameters(),
+                          0.001,
+                          momentum=0.9,
+                          weight_decay=1e-4)
+    if pg.size() > 1:
+        model = torch.nn.parallel.DistributedDataParallel(
+            model,
+            device_ids=[torch.cuda.current_device()],
+            broadcast_buffers=False,
+            process_group=pg)
+
+    measurements = []
+    warmup_iterations = 5
+    measured_iterations = 10
+    for (inputs, target) in (data * (warmup_iterations + measured_iterations)):
+        start = time.time()
+        output = model(*inputs)
+        loss = criterion(output, target)
+        loss.backward()
+        optimizer.step()
+        torch.cuda.synchronize()
+        measurements.append(time.time() - start)
+
+    # Throw away measurements for warmup iterations
+    return measurements[warmup_iterations:]
+
+
+def run_benchmark(prefix, ranks, benchmark):
+    def local_print(msg):
+        if dist.get_rank() == 0:
+            print(msg, end='', flush=True)
+
+    def print_measurements(prefix, nelem, measurements):
+        measurements = sorted(measurements)
+        local_print("%s: " % prefix)
+        for p in [50, 75, 90, 95]:
+            v = np.percentile(measurements, p)
+            local_print("p%02d:  %1.3f  %4d/s  " % (p, v, nelem / v))
+        local_print("\n")
+
+    group_nccl = dist.new_group(ranks=ranks, backend="nccl")
+    group_gloo = dist.new_group(ranks=ranks, backend="gloo")
```

Comment:
Why do we have group_gloo here? Also, it seems like the backend should be configurable via a command line arg to this script.

Reply:
Good point. I've made it configurable.